src/replica/IngestRequestMgr.cc

```diff
@@ -261,14 +343,35 @@ TransactionContribInfo IngestRequestMgr::cancel(unsigned int id) {
     throw IngestRequestNotFound(context_ + string(__func__) + " request " + to_string(id) + " was not found");
 }
 
-IngestRequest::Ptr IngestRequestMgr::next() {
+shared_ptr<IngestRequest> IngestRequestMgr::next() {
     unique_lock<mutex> lock(_mtx);
-    if (_input.empty()) {
-        _cv.wait(lock, [&]() { return !_input.empty(); });
+    shared_ptr<IngestRequest> request;
+    if (request = _next(lock); request == nullptr) {
```

Comment:
My understanding is that as of C++17, the standard supports "if with the initializer", where the following syntax is supported:
```
if (<initializer-statement	>; <condition-to-be-evaluated-for-truth>) {
```
According to the standard:
```
init-statement:
- either an expression statement (which may be a null statement ";")
- a simple declaration, typically a declaration of a variable with an initializer, but it may declare arbitrary many variables or be a structured binding declaration
...
```
See: https://en.cppreference.com/w/cpp/language/if

Based on my understanding of the new syntax,  I thought I could either do:
```
if (shared_ptr<IngestRequest> request = _next(lock); request == nullptr) {
```
or put an expression (the first option in the above-explained syntax) like I did, by:
```
shared_ptr<IngestRequest> request;
if (request = _next(lock); request == nullptr) {
```

Reply:
I chose the last syntax since I'm using the non-`nullptr` reference later, after the `if` block. And I'm effectively using the `if` block for intercepting the special condition (of the null pointer).

Reply:
My idea for init-stmt in `if` is that it is useful mostly for declaring variables within the scope of `if`. Your code does not limit `request` scope to `if`, so it is naturally declared outside `if`. I also think that declaration and initialization should happen together if possible. Using fancy new `if` syntax does not do anything useful in this case.

Reply:
I don't see much difference between this:
```
shared_ptr<IngestRequest> IngestRequestMgr::next() {
    unique_lock<mutex> lock(_mtx);
    shared_ptr<IngestRequest> request;
    if (request = _next(lock); request == nullptr) {
        _cv.wait(lock, [&]() {
            // The mutex is guaranteed to be re-locked here.
            request = _next(lock);
            return request != nullptr;
        });
    }
    return request;
}
```
And this:
```
shared_ptr<IngestRequest> IngestRequestMgr::next() {
    unique_lock<mutex> lock(_mtx);
    shared_ptr<IngestRequest> request = _next(lock); 
    if (request == nullptr) {
        _cv.wait(lock, [&]() {
            // The mutex is guaranteed to be re-locked here.
            request = _next(lock);
            return request != nullptr;
        });
    }
    return request;
}
```

Reply:
Or, it could be:
```
shared_ptr<IngestRequest> IngestRequestMgr::next() {
    unique_lock<mutex> lock(_mtx);
    if (auto request = _next(lock); request != nullptr) {
        return request;
    } else {
        _cv.wait(lock, [&]() {
            // The mutex is guaranteed to be re-locked here.
            request = _next(lock);
            return request != nullptr;
        });
        return request;
    }
}
```
Either version of the code is observable and understandable.

Reply:
if you like the second version I can modify the code accordingly.

Reply:
I like it better because I need to think less about your intentions when I read the code. As it is written now, it made me stop and think about why do you want to do it this way.

Reply:
You have convinced me :)